### ISSUE TITLE
Result sets not closing in 3.0.1?

### ISSUE THREAD

**Issue by leonwang0** [orientechnologies/orientdb#8282]
## Expected behavior  

Is it possible OResultSets in 3.0.1 aren't always released when you call .close()? I believe I am calling it on every result set created, yet am still blowing up:

## Actual behavior  
2018-05-22 16:02:30:765 WARNI {db=sna_new} This database instance has 244850 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:766 WARNI {db=sna_new} This database instance has 262530 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:766 WARNI {db=sna_new} This database instance has 229410 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:766 WARNI {db=sna_new} This database instance has 266900 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 234910 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 220480 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 247970 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 256160 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 244250 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:767 WARNI {db=sna_new} This database instance has 247840 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:768 WARNI {db=sna_new} This database instance has 248810 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:768 WARNI {db=sna_new} This database instance has 251840 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:768 WARNI {db=sna_new} This database instance has 250910 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:769 WARNI {db=sna_new} This database instance has 261970 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:30:769 WARNI {db=sna_new} This database instance has 254890 open command/query result sets, please make sure you close them with OResultSet.close()
2018-05-22 16:02:59:447 WARNI {db=sna_new} This database instance has 252990 open command/query result sets, please make sure you close them with OResultSet.close()^C

## Steps to reproduce  
<add here>

**Comment by luigidellaquila** [comment]
Hi @leonwang0 

Yes, all the result sets are released when you execute .close().
This could be a bug, or just a query you forgot to close.
Do you have a test case to reproduce the problem?

Thanks

Luigi

**Comment by asadganiev** [comment]
I have the same problem. I am using OrientDB with Tinkerpop Gremlin.

**Comment by maggiolo00** [comment]
Hi @asadganiev 

are you running the queries against the GremlinServer?

Thanks

**Comment by asadganiev** [comment]
No. against OrientDB in maven project. Here are project dependencies:

```
<dependencies>
    <dependency>
        <groupId>com.orientechnologies</groupId>
        <artifactId>orientdb-core</artifactId>
        <version>3.0.1</version>
    </dependency>
    <dependency>
        <groupId>com.orientechnologies</groupId>
        <artifactId>orientdb-graphdb</artifactId>
        <version>3.0.1</version>
    </dependency>
    <dependency>
        <groupId>org.apache.tinkerpop</groupId>
        <artifactId>gremlin-core</artifactId>
        <version>3.3.3</version>
    </dependency>
    <dependency>
        <groupId>org.apache.tinkerpop</groupId>
        <artifactId>tinkergraph-gremlin</artifactId>
        <version>3.3.3</version>
    </dependency>
    <dependency>
        <groupId>com.orientechnologies</groupId>
        <artifactId>orientdb-gremlin</artifactId>
        <version>3.0.1</version>
    </dependency>
</dependencies>
```

**Comment by CyprienGottstein** [comment]
I also have this problem on OrientDB 3.0.2 but i must admit i've never seen anywhere that we had to manually close those ResultSet before.

**Comment by asadganiev** [comment]
@GrayDelacluyse 

Are you using Gremlin API too?

**Comment by CyprienGottstein** [comment]
@asadganiev 

No, i'm only using the java Graph API :s

**Comment by asadganiev** [comment]
@tglman 

Thanks, but how can I include it to my maven project dependencies?

**Comment by tglman** [comment]
Hi @asadganiev,

This was a fix for a specific case when the result set is empty, i would definitely suggest to double check your cond for missing close, this patch will be released with 3.0.3, you can get a snapshot from here: https://oss.sonatype.org/content/repositories/snapshots/com/orientechnologies/orientdb-community/3.0.3-SNAPSHOT/

Regards

**Comment by asadganiev** [comment]
@tglman 

I have added these two amended classes (`ODatabaseDocumentAbstract`, `OLocalResultSetLifecycleDecorator`) from this commit [e6de82f](https://github.com/orientechnologies/orientdb/commit/e6de82fa258dbea54888baf75fc961a62a6ffb6e) manually to my project for checking if it is really solves the problem. But unfortunately the issue still exists. 

Later on I am going to create a gist to show a real example of code, so that you can have detailed information.

Best regards,
Asad

**Comment by asadganiev** [comment]
@tglman 
@luigidellaquila 

I have created repository [ODB-8282](https://github.com/asadganiev/ODB-8282) in github.
You check why this issue occurring in my code.

If you any question please write me.

Best regards,
Asad

**Comment by tglman** [comment]
Hi @asadganiev,

we use a lot of sql in the main engine, but not only, I direct load of a record, the save, the commit of transaction, have direct call in the protocol with specific messages.

Regards

**Comment by asadganiev** [comment]
Hi @tglman,

Let me explain why I have asked this question and how I came to OrientDB or even 
to Graph databases.

First I have used Neo4J. It was very slow in insertion. Then found out about OrientDB, it was
fast enough in insertion. But..

As a java developer firstly I chose Graph API. But it was not so easy in using. I was using
Tinkerpop Blueprints, but in github it is shown as deprecated. Then I switched to Tinkerpop3
Gremlin. I liked it very much. Unfortunately there were some issues. Result set is not closed 
in Gremlin, and by the way why I can't set property type in Gremlin? It is storing long as a 
String, and is hanging while searching. Gremlin handles small database, but when I work with
more than 10 million rows it hangs catastrophically.

So now I decided to work with old known pure SQL through JDBC.

**Comment by tglman** [comment]
Hi @asadganiev,

Yes we are aware of some limitations on the gremlin side and we are working as well with upstream to evolve it to be faster. SQL is a fine choice, i would suggest as well to check the Multi Model API : https://www.orientdb.com/docs/last/java/Java-MultiModel-API.html.

Can we close this issue ?

Regards

**Comment by asadganiev** [comment]
Hi @tglman 

Thanks for information, but I think you have some error in your SSL. When tried to open the https://www.orientdb.com/docs/last/java/Java-MultiModel-API.html Safari is showing this:

<img width="759" alt="screen shot 2018-07-05 at 3 56 12 pm" src="https://user-images.githubusercontent.com/7356751/42319632-03404154-806c-11e8-98a0-dffe1d4150b6.png">

**Comment by tglman** [comment]
hi @asadganiev,

Thank you we are checking that.

Regards

**Comment by asadganiev** [comment]
Hi @tglman 

> Can we close this issue ?

Unfortunately the issue still exists, I have tested it with version 3.0.2 and 3.0.3 as well in
[ODB-8282](https://github.com/asadganiev/ODB-8282).

Best regards,
Asad

**Comment by quanghaisoft** [comment]
This error still appears on version 3.0.28, 3.0.31 when calling the wrong function parameter

**Comment by pontusvision** [comment]
Hi, I'm still seeing this issue in 3.1.12!   @Laa, Please re-open this!

**Comment by tglman** [comment]
Hi,

Do you still see this problem in recent version ? Consider also that this warning can happen if you forget to close the result set yourself.
